cclib/io/qcschemawriter.py

```diff
@@ -0,0 +1,251 @@
+                "validated": True,
+            },
+            "provenance": {
+                "creator": metadata["package"],
+                "version": metadata["package_version"],
+                "routine": f"{self.__module__}.{self.__class__.__qualname__}",
+            },
+            "success": metadata["success"],
+            "error": None,
+            "stdout": None,
+            "stderr": None,
+        }
+
+        # TODO This should be derived from a parsed job type.  It's also not
+        # quite right when considering that many jobs (for example, geometry
+        # optimizations) are actually procedures and not drivers.
+        if hasattr(self.ccdata, "hessian"):
+            driver = "hessian"
+        elif hasattr(self.ccdata, "grads"):
+            driver = "gradient"
+        else:
+            driver = "energy"
+        # TODO what is "properties" for?
+        qcschema_dict["driver"] = driver
+
+        # FIXME
+        # if "keywords" in metadata:
+        #     qcschema_dict["keywords"] = set(metadata["keywords"])
+        # else:
+        qcschema_dict["keywords"] = dict()
+
+        # TODO methods should be an enum
+        if metadata["methods"]:
+            method = metadata["methods"][-1]
+            if method == "DFT":
+                method = metadata["functional"]
+        else:
+            # FIXME
+            method = "HF"
+
+        basis_set_name = metadata["basis_set"].lower()
+
+        qcschema_dict["model"] = {
+            "method": method.lower(),
+            "basis": basis_set_name,
+        }
+
+        scf_total_energy = convertor(self.ccdata.scfenergies[-1], "eV", "hartree")
+        mp2_correlation_energy = None
+        mp2_total_energy = None
+        ccsd_correlation_energy = None
+        ccsd_total_energy = None
+
+        if method == "HF":
+            return_energy = scf_total_energy
+        elif metadata["methods"][-1] == "DFT":
+            return_energy = scf_total_energy
+        elif method == "CCSD":
+            if hasattr(self.ccdata, "mpenergies"):
+                mp2_total_energy = convertor(
+                    self.ccdata.mpenergies[-1][-1], "eV", "hartree"
+                )
+                mp2_correlation_energy = convertor(
+                    self.ccdata.mpenergies[-1][-1] - self.ccdata.scfenergies[-1],
+                    "eV",
+                    "hartree",
+                )
+            ccsd_total_energy = convertor(self.ccdata.ccenergies[-1], "eV", "hartree")
+            ccsd_correlation_energy = convertor(
+                self.ccdata.ccenergies[-1] - self.ccdata.scfenergies[-1],
+                "eV",
+                "hartree",
+            )
+            return_energy = ccsd_total_energy
+        else:
+            raise RuntimeError("Don't know what to do with method {}".format(method))
+
+        scf_dipole_moment = None
+        if hasattr(self.ccdata, "moments"):
+            # FIXME We currently have no easy way to tell if the moments from
```

Comment:
We should open an issue about this. I don't this there is an easy way to fix this. If we move to class based properties maybe we could fix this?

Reply:
Yes. Under the current idea, we would have all the different moments grouped together, but in a true provenance model, we would have all properties associated with an (un)relaxed density as children of said density.